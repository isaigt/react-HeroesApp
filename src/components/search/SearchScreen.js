import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroeCard } from '../heroes/HeroeCard';
import { getHeroebyName } from '../../selectors/getHeroebyName';

export const SearchScreen = ({ history }) => {
    
    const location = useLocation();
    const { q='' } = queryString.parse( location.search );

    const [ { search }, handleInputChange ] = useForm({
        search: q
    });
    
    const heroesFiltrados = useMemo(() => getHeroebyName( q ), [ q ]);
    
    const handleSubmit = (e) => {
        e.preventDefault();
        history.push( `?q=${ search }` );
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>

            <div className="row">
                <div className="col-5" >
                    <h4>Search Form</h4>
                    <hr/>

                    <form onSubmit={ handleSubmit } >
                        <input
                            name="search"
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            value={ search }
                            onChange={ handleInputChange }
                        />

                        <button
                            type="submit"
                            className="btn mt-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        ( q === '' )
                        && <div className="alert alert-info animate__animated animate__fadeIn" >Search for a hero</div>
                    }
                    {
                        ( q !== '' && heroesFiltrados.length === 0 )
                        && <div className="alert alert-danger animate__animated animate__shakeX" >No results for "{ q }"</div>
                    }
                    
                    {
                        heroesFiltrados.map( heroe => (
                            <HeroeCard
                                key={ heroe.id }
                                { ...heroe }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
