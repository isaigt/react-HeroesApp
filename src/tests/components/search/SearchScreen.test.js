import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas del componente <Search />', () => {
    test('debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] } >
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').exists() ).toBeTruthy();
    });

    test('debe de mostrarse a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=flash'] } >
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe( 'flash' );
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe de mostrar un error si no se encuentra el Heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=error'] }>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        ); 

        expect( wrapper.find('.alert-danger').exists() ).toBeTruthy();
    });

    test('debe de llamar el HISTORY.PUSH', () => {

        const history = {
            push: jest.fn()
        }
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=error'] }>
                <Route 
                    path="/search" 
                    component={ () => <SearchScreen history={ history } /> }
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'flash'
            }
        });
        
        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });

        expect( history.push ).toHaveBeenCalledWith( '?q=flash' );
        
    });    
});
