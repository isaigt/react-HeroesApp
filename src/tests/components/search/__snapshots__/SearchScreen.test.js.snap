// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <Search /> debe de mostrarse a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="flash"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3 animate__animated animate__fadeInUp"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Flash"
              className="card-img"
              src="./assets/heroes/dc-flash.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Flash
              </h5>
              <p
                className="card-text"
              >
                Jay Garrick
              </p>
              <p
                className="card-text"
              >
                Jay Garrick, Barry Allen, Wally West, Bart Allen
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Flash Comics #1
                </small>
              </p>
              <a
                href="/heroes/flash"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas del componente <Search /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
        Search for a hero
      </div>
    </div>
  </div>
</div>
`;
